### PR TITLE
DATACMNS-1034 - Introduced API to easily register bidirectional, type-based converters.

### PR DESCRIPTION
**Pull request by olivergierke**
Introduced `ConverterBuilder` which exposes API to register Spring `GenericConverter`s for the use with a store module's `CustomConversions`. This allows simple registration of such converters using Java 8 lambdas.

```java
ConverterAware converters = ConverterBuilder
  .reading(String.class, Long.class, it -> Long.valueOf(it))
  .andWriting(it -> Object::toString);
```

The setup can also be done from the reading side which would just need the method invocations inverted. The resulting `ConverterAware` will expose the registered converters so that they can be easily handed to a `CustomConversions` instance. Partial creation of either reading or writing converters is possible, too with the returned instance then only exposing one of the two converters.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ This README as well as the [reference documentation](http://docs.spring.io/sprin
 
 The main project [website](http://projects.spring.io/spring-data/) contains links to basic project information such as source code, JavaDocs, Issue tracking, etc.
 
-For more detailed questions, please refer to [spring-data on stackoverflow](http://stackoverflow.com/questions/tagged/spring-data). If you are new to Spring as well as to Spring Data, look for information about [Spring projects](https://spring.io/projects). 
+For more detailed questions, please refer to [spring-data on stackoverflow](http://stackoverflow.com/questions/tagged/spring-data). If you are new to Spring as well as to Spring Data, look for information about [Spring projects](https://spring.io/projects).
```
Comment (gregturn) [inline readme.md]: Not the place to suggest moving to asciidoctor?

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	 * Creates a new {@link ReadingConverterBuilder} to produce a converter to read values of the given source (the store
+	 * type) into the given target (the domain type).
+	 * 
+	 * @param source must not be {@literal null}.
+	 * @param target must not be {@literal null}.
+	 * @param function must not be {@literal null}.
+	 * @return
+	 */
+	static <S, T> ReadingConverterBuilder<T, S> reading(Class<T> source, Class<S> target,
+			Function<? super T, ? extends S> function) {
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: `Class<T> source` looks backward (S is source, right?)

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	/**
+	 * Interface to represent an intermediate setup step of {@link ConverterAware} defining a reading converter first.
+	 *
+	 * @author Oliver Gierke
+	 * @since 2.0
+	 */
+	interface ReadingConverterBuilder<T, S> extends ConverterBuilder, ReadingConverterAware {
+
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: Should this be `<S, T>`?

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+		 * 
+		 * @param function must not be {@literal null}.
+		 * @return
+		 */
+		ConverterAware andReading(Function<? super T, ? extends S> function);
+	}
+
+	/**
+	 * A {@link ConverterBuilder} ware of both a reading and writing converter.
+	 *
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: *aware*

File: src/main/java/org/springframework/data/convert/DefaultConverterBuilder.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.convert.ConverterBuilder#getConverters()
+	 */
+	@Override
+	public Set<GenericConverter> getConverters() {
+		return Optionals.toStream(getOptionalReadingConverter(), getOptionalWritingConverter()).collect(Collectors.toSet());
+	}
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/DefaultConverterBuilder.java]: Maybe it's just me, but I almost prefer reading this Java 8 stream as..

```
return Optionals
    .toStream(...)
    .collect(...)
```

Even if this results in an extra blank line to comply. ¯\_(ツ)_/¯

File: src/main/java/org/springframework/data/convert/DefaultConverterBuilder.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		@Override
+		public Set<ConvertiblePair> getConvertibleTypes() {
+			return Collections.singleton(convertiblePair);
+		}
+
+		@WritingConverter
+		private static class Writing<S, T> extends ConfigurableGenericConverter<S, T> {
+
+			public Writing(ConvertiblePair convertiblePair, Function<? super S, ? extends T> function) {
+				super(convertiblePair, function);
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/DefaultConverterBuilder.java]: Is public visibility modifier required?

File: src/main/java/org/springframework/data/convert/DefaultConverterBuilder.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+			}
+		}
+
+		@ReadingConverter
+		private static class Reading<S, T> extends ConfigurableGenericConverter<S, T> {
+
+			public Reading(ConvertiblePair convertiblePair, Function<? super S, ? extends T> function) {
+				super(convertiblePair, function);
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/DefaultConverterBuilder.java]: Public required?

File: src/main/java/org/springframework/data/web/PagedResourcesAssembler.java [hunk 8]
```diff
@@ -129,23 +130,41 @@ public void setForceFirstAndLastRels(boolean forceFirstAndLastRels) {
...
 
 		Assert.notNull(link, "Link must not be null!");
-		return createResource(page, assembler, link);
+
+		return createResource(page, assembler, Optional.of(link));
+	}
+
+	/**
+	 * Creates a {@link PagedResources} with an empt collection {@link EmbeddedWrapper} for the given domain type.
+	 *
```
Comment (gregturn) [inline src/main/java/org/springframework/data/web/PagedResourcesAssembler.java]: *empty*

File: src/main/java/org/springframework/data/web/PagedResourcesAssembler.java [hunk 9]
```diff
@@ -129,23 +130,41 @@ public void setForceFirstAndLastRels(boolean forceFirstAndLastRels) {
...
 	}
 
 	/**
 	 * Creates a {@link PagedResources} with an empt collection {@link EmbeddedWrapper} for the given domain type.
 	 * 
 	 * @param page must not be {@literal null}, content must be empty.
 	 * @param type must not be {@literal null}.
-	 * @param link can be {@literal null}.
+	 * @param link must not be {@literal null}.
 	 * @return
```
Comment (gregturn) [inline src/main/java/org/springframework/data/web/PagedResourcesAssembler.java]: Is this actually the place for specify this, given the assertion isn't here?

File: src/test/java/org/springframework/data/convert/ConverterBuilderUnitTests.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		assertThat(converter.getConvertibleTypes())
+				.containsExactly(new ConvertiblePair(source.getClass(), target.getClass()));
+		assertThat(converter.convert(source, TypeDescriptor.forObject(source), TypeDescriptor.forObject(target)))
+				.isEqualTo(target);
+	}
+
+	private static void assertOnlyConverter(ConverterBuilder builder, Supplier<GenericConverter> supplier) {
+		assertThat(builder.getConverters()).containsExactly(supplier.get());
+	}
+}
```
Comment (gregturn) [inline src/test/java/org/springframework/data/convert/ConverterBuilderUnitTests.java]: This test case is simply beautiful.

**Comment by olivergierke** [comment]
Somehow some bogus commits slipped into this one. Removing.

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java (hunk 2 above)
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: Agreed, I flipped those.

**Comment by olivergierke** [comment]
Tweaked code according to @gregturn's suggestions.

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.Set;
+import java.util.function.Function;
+
+import org.springframework.core.convert.converter.GenericConverter;
+import org.springframework.core.convert.converter.GenericConverter.ConvertiblePair;
+import org.springframework.util.Assert;
+
+/**
+ * API to easily set up {@link GenericConverter} instances using Java 8 lambdas, mostly in bidirectional fashion for
+ * easy registration as custom type converters of the SPring Data mapping subsystem. The registration starts either with
```
Comment (mp911de) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: s/SPring/Spring

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+ * the definition of a reading or writing converter that can then be completed
```
Comment (mp911de) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: Add dot at the end.

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+		Assert.notNull(source, "Source type must not be null!");
+		Assert.notNull(target, "Target type must not be null!");
+		Assert.notNull(function, "Conversion function must not be null!");
+
+		return new DefaultConverterBuilder<>(new ConvertiblePair(source, target), Optional.empty(), Optional.of(function));
+	}
+
+	/**
+	 * Creates a new {@link WritingConverterBuilder} to produce a converter to write values of the given source (the
+	 * domain type type) into the given target (the store type).
```
Comment (mp911de) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: there's 2x type in domain type type

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	 */
+	Set<GenericConverter> getConverters();
+
+	/**
+	 * Exposes a writing converter.
+	 *
+	 * @author Oliver Gierke
+	 * @since 2.0
+	 */
+	public interface WritingConverterAware {
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: Interesting, I though at some point, Eclipse presented them as package protected so I thought the default of public only applied to methods. Rolling this back then.

File: src/main/java/org/springframework/data/convert/ConverterBuilder.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+		ConverterAware andReading(Function<? super T, ? extends S> function);
+	}
+
+	/**
+	 * A {@link ConverterBuilder} aware of both a reading and writing converter.
+	 *
+	 * @author Oliver Gierke
+	 * @since 2.0
+	 */
+	public interface ConverterAware extends ConverterBuilder, ReadingConverterAware, WritingConverterAware {}
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/ConverterBuilder.java]: Naming suggestions? ;)

File: src/main/java/org/springframework/data/convert/CustomConversions.java [hunk 16]
```diff
@@ -149,33 +153,46 @@ public boolean isSimpleType(Class<?> type) {
...
+	 * @param candidate must not be {@literal null}.
+	 * @param conversionService must not be {@literal null}.
+	 */
+	private void registerConverterIn(Object candidate, ConverterRegistry conversionService) {
 
-			if (it instanceof Converter) {
-				conversionService.addConverter(Converter.class.cast(it));
-				added = true;
-			}
+		boolean added = false;
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: At a quick glance, yes. Not at a second. The approach used allows the instance to implement multiple interfaces and thus obtains all `ConvertibleType` instances that implementing multiple ones would expose. An early return would stop that process after the first step.

Not sure this is still needed as I am unsure how likely this is in the first place. I just kept it around to preserve behavioural consistency to the old `CustomConversions`. We could think about removing that functionality, though.

**Pull request closed without merging**